File: data/C++/XMLFormatter.cpp
#include "XMLFormatter.h"
#include "PAMFormatter.h"
#include "Exceptions.h"
#include "Layer.h"
#include "Operation.h"
#include "Selection.h"
#include <regex>
#include "Pixel.h"
#include <sstream>
#include <fstream>

void XMLFormatter::write(ImageInfo p, std::string destination) {

	if (destination.substr(destination.length() - 3) != "xml")
		throw NotAppropriateFormat();

	std::ofstream output(destination);

	if (!output)
		throw FileWontOpen();

	//this is somewhat of a header

	std::string text = "";

	text += "<Height>";
	text += std::to_string(p.first.height);
	text += "</Height>\n";

	text += "<Width>";
	text += std::to_string(p.first.width);
	text += "</Width>\n";

	text += "<Pixels>\n";

	for (auto pix : *p.second) {	//we write pixels in RGBA order

		text += std::to_string(pix.getRed());
		text += " ";
		text += std::to_string(pix.getGreen());
		text += " ";
		text += std::to_string(pix.getBlue());
		text += " ";
		text += std::to_string(pix.getAlpha());
		text += " ";
	}

	output << text;
	output << "\n</Pixel>";

	/*
	output << "<Height>" << p.first.height << "</Height>\n";
	output << "<Width>" << p.first.width << "</Width>\n";
	output << "<Pixels>\n";
	
	for (auto pix : *p.second) {	//we write pixels in RGBA order

		output << pix.getRed() << ' ';
		output << pix.getGreen() << ' ';
		output << pix.getBlue() << ' ';
		output << pix.getAlpha() << ' ';
	}

	output << "\n</Pixel>";
	*/
}

/*
ImageInfo XMLFormatter::read(std::string source) {

	if (source.substr(source.length() - 3) != "xml")
		throw NotAppropriateFormat();

	std::ifstream input(source);

	if (!input)
		throw NonExistentFile();

	int h;
	int w;
	std::list<Pixel>* pixels = new std::list<Pixel>;

	std::smatch result;
	std::regex rxH("<Height>([^<]+)</Height>");
	std::regex rxW("<Width>([^<]+)</Width>");

	std::string textLine;

	std::getline(input, textLine);			//getting height
	std::regex_match(textLine, result ,rxH);
	h = atoi(result.str(1).c_str());

	std::getline(input, textLine);			//getting width
	std::regex_match(textLine, result, rxW);
	w = atoi(result.str(1).c_str());

	std::getline(input, textLine);	//getting line <Pixel>
	
	for (int i = 0; i < w * h; i++) {	//getting pixels

		int r, g, b, a;

		input >> r;
		input.get();
		input >> g;
		input.get();
		input >> b;
		input.get();
		input >> a;
		input.get();

		pixels->push_back(Pixel(r, g, b, a));
	}

	return ImageInfo(HeightWidth(h, w), pixels);
}
*/

void addPixels(std::list<Pixel>* pixels,std::string text) {
	std::stringstream stream(text);

	while (1) {
		int r , g, b , a;
		stream >> r;
		if (!stream)
			break;
		stream >> g;
		stream >> b;
		stream >> a;

		pixels->push_back(Pixel(r, g, b, a));

	}
}

ImageInfo XMLFormatter::read(std::string source) {

	if (source.substr(source.length() - 3) != "xml")
		throw NotAppropriateFormat();

	std::ifstream input(source);

	if (!input)
		throw NonExistentFile();

	int h;
	int w;
	std::list<Pixel>* pixels = new std::list<Pixel>;

	std::smatch result;
	std::regex rxH("<Height>([^<]+)</Height>");
	std::regex rxW("<Width>([^<]+)</Width>");

	std::string textLine;

	std::getline(input, textLine);			//getting height
	std::regex_match(textLine, result, rxH);
	h = atoi(result.str(1).c_str());

	std::getline(input, textLine);			//getting width
	std::regex_match(textLine, result, rxW);
	w = atoi(result.str(1).c_str());

	std::getline(input, textLine);	//getting line <Pixel>

	/*
	for (int i = 0; i < w * h; i++) {	//getting pixels

		int r, g, b, a;

		input >> r;
		input.get();
		input >> g;
		input.get();
		input >> b;
		input.get();
		input >> a;
		input.get();

		pixels->push_back(Pixel(r, g, b, a));
	}*/

	std::string pixText;
	std::getline(input, pixText);

	addPixels(pixels, pixText);

	return ImageInfo(HeightWidth(h, w), pixels);
}

std::string XMLFormatter::writeLayer(Layer* layer) {

	std::string text = "";

	text += "<Layer>\n";

	text += "<active>";
	text += std::to_string(layer->isActive() ? 1 : 0);
	text += "</active>\n";

	text += "<visible>";
	text += std::to_string(layer->isVisible() ? 1 : 0);
	text += "</visible>\n";

	text += "<transparency>";
	text += std::to_string((int)(layer->getTransparency() * 100));	//we scale to integer 0-100
	text += "</transparency>\n";

	/*
	text += "<height>";
	text += std::to_string(layer->getHeight());
	text += "</height>\n";

	text += "<width>";
	text += std::to_string(layer->getWidth());
	text += "</width>\n";

	text += "<id>";
	text += std::to_string(layer->getId());
	text += "</id>\n";
	*/

	std::string fileName = "layer";
	fileName += std::to_string(layer->getId());
	fileName += ".xml";
	write(layer->image, fileName);

	text += "<path>";
	text += fileName;
	text += "</path>\n";

	text += "</Layer>\n";

	return text;
}

std::string XMLFormatter::writeSelection(Selection* select) {

	std::string text = "";

	text += "<Selection>\n";

	text += "<name>" + select->getName() + "</name>\n";

	text += "<active>";
	text += std::to_string(select->isActive() ? 1 : 0);
	text += "</active>\n";

	std::string fileName = select->getName() + ".xml";

	std::ofstream output(fileName, std::ofstream::out | std::ofstream::trunc);

	if (!output)
		throw FileWontOpen();

	int i = 0;
	std::string textToWrite = "";

	for (auto point : select->collection) {
		textToWrite += std::to_string(point.getX());
		textToWrite += ",";
		textToWrite += std::to_string(point.getY());
		i++;
		if (i < select->collection.size())
			textToWrite += " ";
	}

	output << textToWrite;

	text += "<Points>" + fileName + "</Points>\n";

	text += "</Selection>\n";

	return text;
}


std::string XMLFormatter::writeOperation(Operation* op) {

	return op->writeXML();
}

File: obradaSlike/XMLFormatter.java
package obradaSlike;

import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class XMLFormatter extends Formatter {

	@Override
	public void write(BufferedImage image, String destination) throws ErrorOccurred {
		
		try {
			FileWriter fileWriter = new FileWriter(destination);
			StringBuilder text = new StringBuilder();
			
			text.append("<Height>" + image.getHeight() + "</Height>\n");
			
			text.append("<Width>" + image.getWidth() + "</Width>\n");
			
			text.append("<Pixels>\n");
			
			for (int i = 0; i < image.getHeight(); i++) 
				for (int j = 0; j < image.getWidth(); j++) {
					Pixel pix = Pixel.makePixel(image.getRGB(j, i));
					
					text.append(pix.getRed() + " ");
					text.append(pix.getGreen() + " ");
					text.append(pix.getBlue() + " ");
					text.append(pix.getAlpha() + " ");
				}
			
			text.append("\n</Pixel>");
			
			fileWriter.write(text.toString());
			
			fileWriter.close();
			
		} catch (IOException e) {
			throw new ErrorOccurred("Error writting file!");
		}
	}

	@Override
	public BufferedImage read(String source) throws ErrorOccurred {
		BufferedImage image = null;
		
		try {
			FileReader fileReader = new FileReader(source);
			BufferedReader buffReader = new BufferedReader(fileReader);
			
			String hText = buffReader.readLine();
			Pattern patH = Pattern.compile("^<Height>([^<]*)</Height>$");
			Matcher matcherH = patH.matcher(hText);
			
			String wText = buffReader.readLine();
			Pattern patW = Pattern.compile("^<Width>([^<]*)</Width>$");
			Matcher matcherW = patW.matcher(wText);
			
			matcherW.matches();
			matcherH.matches();
			
			int h = Integer.parseInt(matcherH.group(1));
			int w = Integer.parseInt(matcherW.group(1));
			
			buffReader.readLine();	//<Pixels>
			
			String pixText = buffReader.readLine();
			
			String[] components = pixText.split(" ");
			
			image = new BufferedImage(w,h,BufferedImage.TYPE_INT_ARGB);
			
			for (int i = 0; i < components.length;) {
				int index = i / 4;
				
				int r = Integer.parseInt(components[i++]);
				int g = Integer.parseInt(components[i++]);
				int b = Integer.parseInt(components[i++]);
				int a = Integer.parseInt(components[i++]);
				
				image.setRGB(index % w, index / w, Pixel.makeARGB(new Pixel(r,g,b,a)));
			}

			buffReader.close();
			
		} catch (IOException e) {
			throw new ErrorOccurred("Error reading file!");
		}
		
		return image;
	}
	
	public String writeLayer(Layer layer) {
		StringBuilder text = new StringBuilder();
		
		text.append("<Layer>\n");
		
		text.append("<active>" + (layer.isActive() ? 1 : 0) + "</active>\n");
		
		text.append("<visible>" + (layer.isVisible() ? 1 : 0) + "</visible>\n");
		
		text.append("<transparency>" + ((int)(layer.getTransparency() * 100)) + "</transparency>\n");
		
		String fileName = "layer" + layer.getId() + ".xml";
		
		try {
			write(layer.getImage(),"data\\" + fileName);
		} catch (ErrorOccurred e) {}
		
		text.append("<path>" + "data\\" + fileName + "</path>\n");
		
		text.append("</Layer>\n");
		
		return text.toString();
	}
	
	public String writeSelection(Selection selection) {
		StringBuilder text = new StringBuilder();
		
		text.append("<Selection>\n");
		
		text.append("<name>" + selection.getName() + "</name>\n");
		
		text.append("<active>" + (selection.isActive() ? 1 : 0) + "</active>\n");
		
		String fileName = selection.getName() + ".xml";
		
		try {
			FileWriter fileWriter = new FileWriter("data\\" + fileName);
			StringBuilder pointsText = new StringBuilder();
			
			selection.getAllPoints().stream().forEach(p -> {
				pointsText.append(p.getX() + "," + p.getY() + " ");		
			});
			
			if (pointsText.length() > 1)
				pointsText.deleteCharAt(pointsText.length() - 1);
			
			fileWriter.write(pointsText.toString());	
			fileWriter.close();
			
		} catch (IOException e) {}
		
		text.append("<Points>" + "data\\" + fileName + "</Points>\n");	//we add data\ , because we save it in data folder
		
		text.append("</Selection>\n");
		
		return text.toString();
	}
	
	public String writeOperation(Operation operation) {
		
		return operation.writeXML();
	}

}
